UCP/PROTO: Filter lanes according to distance from GPU (proto v2)

shasson5 opened the pull request:
## What
Filter lanes according to distance from GPU

## Why ?
Improve performance for GPU machines where multi lane protocols are involved

yosefe commented on src/ucp/proto/proto_multi.c:
```diff
@@ -84,6 +84,11 @@ ucs_status_t ucp_proto_multi_init(const ucp_proto_multi_init_params_t *params,
     lane_map                = 0;
     max_frag_ratio          = 0;
     for (i = 0; i < num_lanes; ++i) {
+        if (ucs_popcount(lane_map) == params->max_lanes) {
+            ucs_trace("reached max lanes");
```
1. add assert that ucs_popcount(lane_map) <= params->max_lanes
2. IMO better remove this ucs_trace

gleon99 replied:
Or maybe add `&& ucs_popcount(lane_map) < params->max_lanes` inside the `for` condition instead?

shasson5 replied:
instead of assert?

gleon99 replied:
@yosefe  Why put an `assert` here? We only need a stop condition for the loop

yosefe replied:
assert that it's always <=
since we stop at ==

yosefe commented on src/ucp/wireup/select.c:
```diff
@@ -1685,7 +1685,9 @@ ucp_wireup_add_rma_bw_lanes(const ucp_wireup_select_params_t *select_params,
-    bw_info.max_lanes                 = context->config.ext.max_rndv_lanes;
+    bw_info.max_lanes                 = context->config.ext.proto_enable ?
+                                        UCP_MAX_LANES :
+                                        context->config.ext.max_rndv_lanes;
```
this may create too many uct endpoints (lanes)
let's at least filter out those who lave low bandwidth compared to others (according to the configured threshold)?
going forward, we can connect the slow-path lanes on demand and not immediately during ucp endpoint creation.

brminich replied:
looks like it may break wire compatibility

shasson5 replied:
That's correct but for now we decided to continue with this solution, because this task is a bit urgent.
Will be fixed later on.

brminich replied:
imo, there is no easy way to fix it

yosefe replied:
maybe we could pass some flag in the address to indicate max lanes?
or connect some lanes later on if there is a difference between client/server (handle "reconfig") - but that's not simple ..

gleon99 commented on src/uct/base/uct_iface.c:
```diff
@@ -857,6 +857,92 @@ void uct_iface_get_local_address(uct_iface_local_addr_ns_t *addr_ns,
     }
 }
 
+void uct_iface_set_sys_dev(const char *dev_name, const char *sysfs_path,
```
Maybe move this func and the next one to UCS?

shasson5 replied:
won't it complicate topo API?

yosefe replied:
IMO also it should be in ucs/topo since it deals with sysfs paths

yosefe replied:
return sys_dev instead of void

gleon99 commented on src/uct/base/uct_iface.h:
```diff
@@ -865,6 +865,32 @@ void uct_iface_get_local_address(uct_iface_local_addr_ns_t *addr_ns,
+ *                             if it already exists).
+ * @param [out] sys_dev        Holds a topo module identifier for the device.
+ */
+void uct_iface_set_sys_dev(const char *dev_name, const char *sysfs_path,
```
Maybe put is_primary last (a flag)

shasson5 replied:
after the output argument?

shasson5 commented on src/ucp/core/ucp_context.c:
```diff
@@ -908,6 +908,21 @@ static int ucp_is_resource_enabled(const uct_tl_resource_desc_t *resource,
+        return 0;
+    }
+
+    return (resource1->sys_device == resource2->sys_device);
```
I think it would be a problem because this function compare different types of transports.

yosefe commented on src/ucp/proto/proto_multi.c:
```diff
@@ -55,7 +55,7 @@ ucs_status_t ucp_proto_multi_init(const ucp_proto_multi_init_params_t *params,
                                              params->middle.lane_type,
                                              params->middle.tl_cap_flags,
-                                             params->max_lanes - 1,
+                                             UCP_MAX_LANES - 1,
```
UCP_PROTO_MAX_LANES

yosefe commented on src/ucp/wireup/select.c:
```diff
@@ -1415,6 +1415,13 @@ ucp_wireup_am_bw_score_func(const ucp_worker_iface_t *wiface,
+static int
+ucp_wireup_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
+{
+    return (context->config.ext.proto_enable ||
```
extra ()

yosefe commented on src/ucp/wireup/select.c:
```diff
@@ -1685,7 +1692,9 @@ ucp_wireup_add_rma_bw_lanes(const ucp_wireup_select_params_t *select_params,
     bw_info.criteria.lane_type        = UCP_LANE_TYPE_RMA_BW;
-    bw_info.max_lanes                 = context->config.ext.max_rndv_lanes;
+    bw_info.max_lanes                 = context->config.ext.proto_enable ?
+                                        UCP_MAX_LANES :
```
maybe UCP_PROTO_MAX_LANES

yosefe commented on src/ucs/sys/topo/base/topo.c:
```diff
@@ -324,6 +324,49 @@ const char *ucs_topo_distance_str(const ucs_sys_dev_distance_t *distance,
     return ucs_string_buffer_cstr(&strb);
 }
 
+ucs_sys_device_t ucs_topo_set_sys_dev(const char *dev_name,
```
ucs_topo_get_sysfs_dev

yosefe commented on src/ucs/sys/topo/base/topo.c:
```diff
@@ -324,6 +324,49 @@ const char *ucs_topo_distance_str(const ucs_sys_dev_distance_t *distance,
+
+    /* Override device name if device is new (number of devices increased),
+       or if it's a primary device */
+    override_name = (num_devices_pre < ucs_topo_num_devices()) || is_primary;
```
IMO ucs_topo_num_devices check is not needed
what if just pass is_primary as override_name?

shasson5 replied:
what about TCP devices which are non-IB? (eg. ens32)
They will not get a name at all.

yosefe replied:
Why not? if the device is new, it will use ens32 (or ib0 for example)
if mlx* device comes along with same id - it will override.
if mlx* device comes first, then ib0 will have override_name=0 so mlx* name will stay.

yosefe commented on src/ucs/sys/topo/base/topo.c:
```diff
@@ -324,6 +324,49 @@ const char *ucs_topo_distance_str(const ucs_sys_dev_distance_t *distance,
+    override_name = (num_devices_pre < ucs_topo_num_devices()) || is_primary;
+    status = ucs_topo_sys_device_set_name(sys_dev, dev_name, override_name);
+
+    ucs_assert_always(status == UCS_OK);
```
if failed return unknown

shasson5 replied:
won't it cause a corruption in `ucs_topo_global_ctx.devices` ?

yosefe replied:
it's created with a default name.. so now i understand why passing override_name directly is not good enough
maybe we should keep a field in ucs_topo_global_ctx.devices (per device) that means the "priority" of the name.. and compare it with a priority argument provided to ucs_topo_sys_device_set_name()

yosefe commented on src/ucs/sys/topo/base/topo.c:
```diff
@@ -379,9 +422,11 @@ ucs_topo_sys_device_set_name(ucs_sys_device_t sys_dev, const char *name)
-                                                           "sys_dev_name");
+    if (override_name) {
+        ucs_free(ucs_topo_global_ctx.devices[sys_dev].name);
+        ucs_topo_global_ctx.devices[sys_dev].name = ucs_strdup(name,
+                                                               "sys_dev_name");
```
line seems too long (pls check clang fmt)

shasson5 replied:
it's exactly 80 chars, clang does not warn about it.

yosefe commented on src/ucs/sys/topo/base/topo.h:
```diff
@@ -138,6 +138,21 @@ ucs_status_t ucs_topo_get_distance(ucs_sys_device_t device1,
+ *
+ * @param [in]  dev_name       Device Name.
+ * @param [in]  sysfs_path     sysfs path for the required device.
+ * @param [in]  is_primary     Indicates whether to treat the device as
```
maybe rename is_primary to overide_name?

yosefe commented on src/ucs/sys/topo/base/topo.c:
```diff
@@ -31,8 +31,9 @@
-    char             *name;
+    ucs_sys_bus_id_t                bus_id;
+    char                            *name;
+    ucs_topo_device_name_priority_t name_priority;
```
this could be just an "unsigned" type with numbers
default (bdf name) is 0
TCP will pass 10
IB will pass 20

yosefe commented on src/ucs/sys/topo/base/topo.c:
```diff
@@ -344,20 +345,15 @@ ucs_sys_device_t ucs_topo_set_sys_dev(const char *dev_name,
-    ucs_assert_always(status == UCS_OK);
+    status = ucs_topo_sys_device_set_name(sys_dev, dev_name, name_priority);
+    if (status != UCS_OK) {
+        goto out_unknown;
```
actually if failed we can still return sys_dev

yosefe commented:
/azp run

yosefe commented on src/ucs/sys/topo/base/topo.c:
```diff
@@ -172,8 +172,7 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
         ucs_topo_global_ctx.devices[*sys_dev].name   = name;
-        ucs_topo_global_ctx.devices[*sys_dev].name_priority =
-                UCS_TOPO_DEVICE_NAME_PRIORITY_DEFAULT;
+        ucs_topo_global_ctx.devices[*sys_dev].name_priority = 0;
```
align = on column

yosefe commented:
@shasson5 how did you test that more lanes are actually created now?

shasson5 commented:
1) I ran the following command on a DGX machine:
```
mpirun --host swx-dgx03:2 --output-filename log --display-map -np 2 --map-by ppr:2:node -mca pml ucx -mca btl self -x CUDA_VISIBLE_DEVICES=4,5  -x UCX_TLS=ib,cuda_copy -x  UCX_MAX_RNDV_LANES=1  -x UCX_RNDV_SCHEME=get_zcopy -x UCX_PROTO_ENABLE=y -x LD_LIBRARY_PATH=/.autodirect/mtrswgwork/shasson/ucx/install/lib/:$LD_LIBRARY_PATH $HPCX_OSU_CUDA_DIR/osu_bw -d cuda D D
```
2) I verified more lanes were created by looking at the log file:
```
[1667208003.244345] [swx-dgx04:1730 :0]    proto_select.c:255  UCX  TRACE       trying rndv/get/zcopy
[1667208003.244351] [swx-dgx04:1730 :0]    proto_common.c:406  UCX  TRACE         selecting up to 1/4 lanes for read from remote rndv_recv() into cuda/GPU1 from cuda/dev[1]
[1667208003.244355] [swx-dgx04:1730 :0]    proto_common.c:521  UCX  TRACE           lane[0] rc_mlx5/mlx5_0:1: added as lane 0
[1667208003.244359] [swx-dgx04:1730 :0]    proto_common.c:526  UCX  TRACE           selected 1 lanes
[1667208003.244364] [swx-dgx04:1730 :0]    proto_common.c:406  UCX  TRACE         selecting up to 15/4 lanes for read from remote rndv_recv() into cuda/GPU1 from cuda/dev[1]
[1667208003.244368] [swx-dgx04:1730 :0]    proto_common.c:521  UCX  TRACE           lane[1] rc_mlx5/mlx5_1:1: added as lane 1
[1667208003.244371] [swx-dgx04:1730 :0]    proto_common.c:521  UCX  TRACE           lane[2] rc_mlx5/mlx5_2:1: added as lane 2
[1667208003.244455] [swx-dgx04:1730 :0]    proto_common.c:521  UCX  TRACE           lane[3] rc_mlx5/mlx5_3:1: added as lane 3
[1667208003.244459] [swx-dgx04:1730 :0]    proto_common.c:526  UCX  TRACE           selected 3 lanes
```
3) I verified that the correct lane was indeed selected (mlx5_2):
`[1667208003.244656] [swx-dgx04:1730 :0]     proto_multi.c:106  UCX  TRACE         lane[2] send_pre_overhead: 40.00 ns send_post_overhead: 60.00 ns recv_overhead: 40.00 ns bw 5897.12 latency: 1172.00 ns`

gleon99 commented:
@shasson5  consider rephrasing the commit + PR title.
"Lanes are filtered..." --> "Filter lanes..."  i.e. have some "action" that his change creates.

yosefe commented:
also the CI fails consistently with this PR, so it can be real issue

shasson5 commented on src/ucs/sys/topo/base/topo.c:
```diff
@@ -164,7 +164,7 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
 
         /* Set default name to abbreviated BDF */
-        name = ucs_malloc(UCS_SYS_BDF_NAME_MAX, "sys_dev_bdf_name");
+        name = ucs_calloc(UCS_SYS_BDF_NAME_MAX, 1, "sys_dev_bdf_name");
```
revereted

gleon99 commented:
@shasson5  also please rephrase the PR description.

gleon99 commented on src/ucp/wireup/select.c:
```diff
@@ -1415,6 +1415,13 @@ ucp_wireup_am_bw_score_func(const ucp_worker_iface_t *wiface,
 }
 
+static int
+ucp_wireup_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
```
As it is a predicate, I'd add `is` to the name (like `uct_ib_md_is_flush_rkey_valid` for example).

gleon99 commented on src/ucs/sys/topo/base/topo.h:
```diff
@@ -164,17 +177,19 @@ ucs_topo_find_device_by_bdf_name(const char *name, ucs_sys_device_t *sys_dev);
- * @param [in]  sys_dev  System device to set the name for.
- * @param [in]  name     Name to set for this system device. Note: the name can
- *                       be released after this call.
+ * @param [in]  sys_dev        System device to set the name for.
```
Can remove spaces after arg name

gleon99 commented on src/uct/base/uct_iface.h:
```diff
@@ -865,6 +865,19 @@ void uct_iface_get_local_address(uct_iface_local_addr_ns_t *addr_ns,
+/**
+ * Returns sysfs path for the required device.
+ *
+ * @param [in]  dev_path       Device file path.
```
Can remove excessive spacing

gleon99 commented:
👍🏼 

@yosefe ?

yosefe reviewed (commented):
the code LGTM but CI still fails

yosefe commented on src/ucp/wireup/select.c:
```diff
@@ -384,6 +398,64 @@ ucp_wireup_select_reg_mem_types(ucp_context_h context, ucp_md_index_t md_index)
     return reg_mem_types;
 }
 
+static ucp_wireup_remote_device_count_t *ucp_wireup_get_remote_device_count(
```
the logic of fixing dev_index should be part of address unpack
i would expect select.c code to not change at all

gleon99 commented:
@shasson5 pipeline failure seems relevant.

gleon99 commented on src/ucp/wireup/select.c:
```diff
@@ -60,15 +61,28 @@ typedef struct {
+    unsigned         count;
+    ucp_rsc_index_t  dev_index;
+    ucs_sys_device_t sys_dev;
+} ucp_wireup_remote_device_count_t;
```
Can we try to think of another naming here?
`count_t` which includes `count` is a bit weird.

gleon99 commented on src/ucp/wireup/select.c:
```diff
@@ -407,8 +407,13 @@ static ucp_wireup_remote_device_count_t *ucp_wireup_get_remote_device_count(
     for (i = 0; i < ucs_array_length(count_array); ++i) {
         count_element = &ucs_array_elem(count_array, i);
+        /* We use 'or' in this condition because 2 devices can have different
+         * dev_index (if they have different MDs) but they still considered same
```
`the are still`

gleon99 replied:
`same` -> `the same`

gleon99 commented on src/ucp/wireup/select.c:
```diff
@@ -437,7 +509,8 @@ static UCS_F_NOINLINE ucs_status_t ucp_wireup_select_transport(
         addr_index = ucp_unpacked_address_index(address, ae);
-        if (!(remote_dev_bitmap & UCS_BIT(ae->dev_index))) {
+        if (ucp_wireup_remote_device_count_exceeded(remote_dev_count, ae)) {
+            /* exclude remote device if reached max concurrency level */
```
`Exclude`
`max concurrency level is reached`

shasson5 replied:
reverted

yosefe commented on src/ucp/wireup/address.c:
```diff
@@ -164,6 +165,20 @@ typedef struct {
+
+typedef ucs_array_t(ucp_address_remote_device) ucp_address_remote_device_array_t;
+
+UCS_ARRAY_IMPL(ucp_address_remote_device, unsigned,
```
maybe use UCS_ARRAY_DEFINE_INLINE ?

yosefe commented on src/ucp/wireup/address.c:
```diff
@@ -1469,10 +1484,38 @@ ucs_status_t ucp_address_pack(ucp_worker_h worker, ucp_ep_h ep,
+    ucp_rsc_index_t index;
+    ucp_address_remote_device_t *device;
+
+    for (index = 0; index < ucs_array_length(device_array); ++index) {
```
ucs_array_for_each

shasson5 commented:
/azp run

gleon99 commented on src/ucp/wireup/select.c:
```diff
@@ -1415,6 +1415,13 @@ ucp_wireup_am_bw_score_func(const ucp_worker_iface_t *wiface,
 }
 
+static int
+ucp_wireup_is_md_map_count_valid(ucp_context_h context, ucp_md_map_t md_map)
```
inline

gleon99 commented on src/ucp/core/ucp_worker.inl:
```diff
@@ -219,7 +219,9 @@ ucp_worker_common_address_pack_flags(ucp_worker_h worker)
 
-    if (worker->context->num_mem_type_detect_mds > 0) {
+    if ((worker->context->num_mem_type_detect_mds > 0) ||
+        /* todo: This code breaks wire compatability - fix it */
```
`TODO`

brminich commented on src/ucs/sys/topo/base/topo.h:
```diff
@@ -138,6 +137,20 @@ ucs_status_t ucs_topo_get_distance(ucs_sys_device_t device1,
                                   char *buffer, size_t max);
 
+/**
+ * Sets a system device.
```
`gets` and `sets if not found`?

brminich commented on src/uct/ib/base/ib_device.c:
```diff
@@ -620,9 +539,8 @@ ucs_status_t uct_ib_device_query(uct_ib_device_t *dev,
-                                              path_buffer);
-    uct_ib_device_set_sys_dev(dev, sysfs_path);
+    sysfs_path = uct_iface_get_sysfs_path(dev_path, dev_name, path_buffer);
+    dev->sys_dev = ucs_topo_get_sysfs_dev(dev_name, sysfs_path, 20);
```
pls use some macro definition instead of magic number

brminich commented on src/uct/base/uct_iface.c:
```diff
@@ -857,6 +857,54 @@ void uct_iface_get_local_address(uct_iface_local_addr_ns_t *addr_ns,
+    char *sysfs_path;
+    int ret;
+
+    /* PF: realpath name is of form /sys/devices/.../0000:03:00.0/infiniband/mlx5_0 */
+    /* SF: realpath name is of form /sys/devices/.../0000:03:00.0/<UUID>/infiniband/mlx5_0 */
```
minor: can also mention some net/eth device, as this func is not just for infiniband now

yosefe commented on src/ucp/wireup/address.c:
```diff
@@ -1488,18 +1485,19 @@ static ucp_rsc_index_t ucp_address_get_remote_device_index(
         ucp_rsc_index_t dev_index, ucs_sys_device_t sys_dev)
 {
-    ucp_rsc_index_t index;
+    ucp_rsc_index_t index = 0;
```
index -> dev_index

yosefe commented on src/ucs/sys/topo/base/topo.h:
```diff
@@ -26,6 +26,11 @@ BEGIN_C_DECLS
 #define UCS_SYS_BDF_NAME_MAX 16
 
+/* System device name priority */
+#define UCS_SYS_DEVICE_PRIORITY_TCP  10
+#define UCS_SYS_DEVICE_PRIORITY_IB   20
+#define UCS_SYS_DEVICE_PRIORITY_CUDA 10
```
UCS code cannot be aware of transports

yosefe commented:
@shasson5 pls squash
